src/miner.cpp

```diff
@@ -457,3 +462,283 @@ void IncrementExtraNonce(CBlock* pblock, const CBlockIndex* pindexPrev, unsigned
+
+            std::unique_ptr< BlockAssembler > assembler( new BlockAssembler( chainparams ) ) ;
+            std::unique_ptr< CBlockTemplate > pblocktemplate( assembler->CreateNewBlock( coinbaseScript->reserveScript ) ) ;
+            if ( ! pblocktemplate.get() )
+            {
+                LogPrintf( "Keypool ran out, please invoke keypoolrefill before restarting the mining thread\n" ) ;
+                return ;
+            }
+
+            CBlock * pblock = &pblocktemplate->block ;
+            IncrementExtraNonce( pblock, pindexPrev, nExtraNonce ) ;
+
+            //
+            // Search
+            //
+
+            uint32_t hashesScanned = 0 ;
+            int64_t scanBeginsMillis = GetTimeMillis() ;
+            const CBlockHeader * const blockHeader = pblock ;
+            arith_uint256 solutionHash = arith_uint256().SetCompact( blockHeader->nBits ) ;
+            uint256 hash ;
+            uint32_t nNonce = randomNumber() ;
+
+            LogPrintf(
+                "Running CoinMiner (%c) with %u transactions in block (%u bytes), looking for scrypt hash <= %s, random initial nonce 0x%x\n",
+                threadChar,
+                pblock->vtx.size(),
+                ::GetSerializeSize( *pblock, SER_NETWORK, PROTOCOL_VERSION ),
+                solutionHash.GetHex(), nNonce
+            ) ;
+
+            while ( true )
+            {
+                // Check if something found
+                if ( ScanScryptHash( *blockHeader, nNonce, &hash, solutionHash, hashesScanned ) )
+                {
+                    if ( UintToArith256( hash ) <= solutionHash )
+                    {
+                        // Found a solution
+                        pblock->nNonce = nNonce ;
+                        if ( hash != pblock->GetPoWHash() ) {
+                            LogPrintf( "CoinMiner (%c): oops! ScanScryptHash found %s but block with nonce 0x%x has scrypt hash %s\n",
+                                       threadChar, hash.GetHex(), pblock->nNonce, pblock->GetPoWHash().GetHex() ) ;
+                            throw std::runtime_error( "hash != pblock->GetPoWHash()" );
+                        }
+
+                        LogPrintf( "CoinMiner (%c):\n", threadChar ) ;
+                        LogPrintf( "proof-of-work found with nonce 0x%x\n   scrypt hash %s\n   <= solution %s\n",
+                                   nNonce, hash.GetHex(), solutionHash.GetHex() ) ;
+
+                        /* bool ok = */ ProcessBlockFound( pblock, chainparams ) ;
+                        coinbaseScript->KeepScript() ;
+
+                        // for regression testing, stop mining after a block is found
+                        if ( chainparams.MineBlocksOnDemand() )
+                            throw boost::thread_interrupted() ;
+
+                        break ;
+                    }
+                    else
+                    {
+                        nNonce = randomNumber() ;
```

Comment:
this can really go one level up

Reply:
like I did here https://github.com/VadiqueMe/dogecoin/commit/01679458f065203888e303df75ede061705c539d